src/zkevm_specs/evm_circuit/execution/push.py

```diff
@@ -6,19 +7,40 @@ def push(instruction: Instruction):
     opcode = instruction.opcode_lookup(True)
     num_pushed = opcode - Opcode.PUSH1 + 1
     num_additional_pushed = num_pushed - 1
+    code_length = instruction.bytecode_length(instruction.curr.code_hash)
+    code_length_left = code_length - instruction.curr.program_counter
+    is_out_of_bound, _ = instruction.compare(
+        code_length_left, num_additional_pushed, N_BYTES_PROGRAM_COUNTER
+    )
 
     value = instruction.stack_push()
     value_le_bytes = value.to_le_bytes()
     selectors = instruction.continuous_selectors(num_additional_pushed, 31)
+    counters = [
+        FQ(x)
+        for x in [
+            0
+            if selectors[i] == 0 or i == 0
+            else (
+                (
+                    i < num_additional_pushed.expr().n
+                    and i > FQ(num_additional_pushed - code_length_left).expr().n
+                )
+                if (is_out_of_bound.expr().n == 1)
+                else i < num_additional_pushed.expr().n
+            )
+            for i in range(31)
+        ]
+    ]
 
     for idx in range(32):
         index = instruction.curr.program_counter + num_pushed - idx
-        if idx == 0 or selectors[idx - 1] == 1:
+        if (idx == 0 and not is_out_of_bound) or (idx > 0 and counters[idx - 1] == 1):
             instruction.constrain_equal(
                 value_le_bytes[idx], instruction.opcode_lookup_at(index, False)
             )
-        else:
-            instruction.constrain_zero(value_le_bytes[idx])
+    for idx in range(1, 31):
+        instruction.constrain_zero((1 - counters[idx]) * value_le_bytes[idx + 1])
```

Comment:
Previously `value_le_bytes[idx]` is always constraint by `opcode_lookup_at`, so we could skip the `constrain_zero` for it, but with padding I think we need to check every byte is zero if it's not constraint by `opcode_lookup_at`.

Here it seems `value_le_bytes[0..2]` are always skipped right?

Reply:
That's much better indeed!
I was trying not to increase the degree of the identity but it turned everything more complicated and I even made an error with the values skipped as you saw.

Reply:
Updated!

Reply:
I see I see, I think it makes complete sense to not increase the degree, there must be smarter way to generate `is_pushed and (not is_padding)` I feel. Also in circuit side the constraint builder will help us to cap it at some value by using some extra intermediate cells, so I think it might be okay here to write in this way.